### PR TITLE
DietPi-VPN | Rework from DietPi-NordVPN with ProtonVPN and custom provider support

### LINKED ISSUE #3948: DietPi-Software | ProtonVPN
#### Formal software information
- Software title | ProtonVPN
- Short description | Add protonvpn client and vpn gateway
- Official URL (if available) | https://protonvpn.com/blog/linux-vpn-cli-beta/
- (Official) install documentation | https://protonvpn.com/blog/linux-vpn-cli-beta/

#### Are there similar/alternative software titles available with DietPi-Software?
NordVPN
- ...

#### What makes your requested software better than the above solutions, if available?

- the vpn should also be the gateway like this:
https://gist.github.com/superjamie/ac55b6d2c080582a3e64

#### How can DietPi make the installation easier or compatible, than following the install instructions or do APT installation, if available?

- ...

#### Can you provide the installation steps that you would suggest DietPi-Software to do?

1. ...

#### Are you willing to help maintaining the software installation, e.g. in case of needed setup changes due to updates etc.? This is not needed, but could speed up our decision to implement it, as man power is always a topic :wink:.

- ...

### PR DESCRIPTION
**Pull request by nonnorm**
**Status**: Ready
- [x] Add way to connect to ProtonVPN
- [x] Add way to just use generic OpenVPN file
- [x] Test with ProtonVPN
- [x] Test with NordVPN
- [x] Add killswitch

**Reference**: https://github.com/MichaIng/DietPi/issues/3948

**Commit list/description**:

- DietPi-VPN | Start transition from dietpi-nordvpn to dietpi-vpn
- DietPi-Software | Add way to download and parse ProtonVPN server list
- DietPi-VPN | Add way to select ProtonVPN Servers
- DietPi-VPN | Add way to generate .ovpn file
- DietPi-Software | Switch from NordVPN to VPN
- DietPi.bash | Add missing DietPi-VPN alias
- Patches | Remove DietPi-NordVPN

### CODE REVIEW

**Comment by nonnorm** [comment]
I'm a little lost on why the `G_WHIP_MENU` to select a server isn't working.  The array seems to be fine, and it worked when I didn't remove the first line.
**EDIT**: The array doesn't actually work outside of the loop for some reason.  Inside everything's fine.

**Comment by MichaIng** [comment]
Awesome! I didn't know about and love `envsubst` for automatic variable call substitutions in files.

I'll test what is wrong with the array, at least I cannot see an obvious issue currently. Pipes and redirects work a bit different, probably you can try:
```
while read -r server
do
	G_WHIP_MENU_ARRAY+=("$(echo "$server" | awk -F\" '/"Name":/{print $4;exit}')" '')
done < <(tail -n +2 "$PROTONVPN_SERVERS")
```
to bring the list into the loop. Also the list could be trimmed in the first place, so that it's much easier and faster to create the menu from that list:
```
curl -sSfL https://api.protonvpn.ch/vpn/logicals | sed 's/"Domain":"/\n/g' | mawk -F\" 'NR % 2 == 0 {print $1}'
```
- Add newline before each domain.
- Print only until first `"`, which contains the domain.
- Print only even lines, since there are two identical domain fields for each server and to exclude the first line.
___________
A few other ideas:
- Since we have an "update" feature which downloads the server list/configs, actually the `dietpi-software` install option is obsolete IMO. We should simply remove it. But we need to add a provider selection menu entry. When `dietpi-vpn` is called the first time, one needs to select the VPN provider, which downloads/updates the list/configs for the chosen provider. The steps for the "Update" entry need to become an own function then.
- The ProtonVPN OVPN file creation should btw be done as part of the `Save_Settings` function, so that selections (aside of provider selection of course) stay purely temporary and do nothing until "Apply" is selected.
- To not have OVPN file overlaps, sub directories for each providers would be an idea, which then also serves as indicator whether the server list does already exist or not for the chosen provider.
- As currently all ProtonVPN configs are removed, before a new one is created, what about keeping it simple and create a single `protonvpn.ovpn` file that is simply overwritten? No need to keep the domain/server name and port part of the config file name, IMO. In case of NordVPN, it's simply the file names their archive ships, so for simplicity that is kept. Although in case of NordVPN,  the chosen servers OVPN could be copied to a static `nordvpn.ovpn` as well. ... Not thinking about it, why not using a single `client.ovpn` file for all providers, so the service file does not even need to be changed and we can make it a static service file here: https://github.com/MichaIng/DietPi/tree/master/rootfs/etc/systemd/system ... So `/etc/openvpn/nordvpn/` and `/etc/openvpn/protonvpn/` for provider-specific stuff and `/etc/openvpn/client.ovpn` is the config that is always used.

**Comment by nonnorm** [comment]
Well, now it shows I'm connected, but no traffic seems to be going through it (thus my IP address stays the same).
**EDIT**: Actually, traffic is going through it, but every site is telling me that my IP address is still the same.

**Comment by MichaIng** [comment]
Awesome work, I live the feature to add an own custom `.ovpn` file, giving it a generic usage ability ❤️. Is it ready from your side for testing? I have a NordVPN account at least to run tests with.
___________
We should btw migrate existing NordVPN configs, i.e. read the chosen NordVPN server from the existing config file, copy it to `/etc/openvpn/client.ovpn`, check whether the old service is enabled, and in case disable it + enable the new one. So no running connection is aborted and on reboot the same `.ovpn` is loaded by the new service.

**Comment by nonnorm** [comment]
Not quite, when using ProtonVPN it shows that it's connected (and indeed data seems to be going through the `tun0` interface), but my IP address is staying the same.

**Comment by MichaIng** [comment]
I compared the ProtonVPN template with the NordVPN configs, and they match completely, aside of:
```
ping 15
ping-restart 0
ping-timer-rem
```
being additionally present in NordVPN configs. It's to send a ping (not ICMP, but same purpose) to the server when for 15 seconds no other packet has been send, to check the connection state, especially useful for UDP connections. When no answer is received for X seconds, a restart is scheduled. By default X=120, but `ping-restart 0` unsets this default "until the client pulls a replacement value from the server". `ping-timer-rem` adds the condition that a remote must be present for the ping-restart to happen, but that is obsolete here since the config has a fixed remote. Since ProtonVPN configs do not have the ping set, I guess one does not know the connection state, i.e. the daemon won't error out when the connection is lost until an actual package is tried to be send.

The matter is discussed here: https://github.com/ProtonVPN/linux-cli/issues/28
Interestingly I can't find someone mentioning this OpenVPN native ping/restart feature. But again interestingly many log posts on the repo show that `ping 10,ping-restart 60` is set 🤔. Probably this was set previously or is set elsewhere outside of the config template?

And the other difference is the additional remote entries with other ports. With `remote-random` a random one is used, but is it reasonable to randomly use port 80 and 443 as well here?
______________
But @ravenclaw900 let's check your connection issue:
- Essential for packets being send through the VPN is two routes being created by the OpenVPN daemon. Can you check: `ip r`
- If the routes are indeed missing, does the daemon log any errors? `journalctl -u dietpi-vpn`

**Comment by nonnorm** [comment]
Routes seem to be present:
```
root@DietPiVBOX:~# ip r
0.0.0.0/1 via 10.19.0.1 dev tun0 
default via 192.168.1.1 dev eth0 
10.19.0.0/16 dev tun0 proto kernel scope link src 10.19.0.12 
89.39.107.190 via 192.168.1.1 dev eth0 
128.0.0.0/1 via 10.19.0.1 dev tun0 
192.168.1.0/24 dev eth0 proto kernel scope link src 192.168.1.189 
```
And no obvious errors:
```
root@DietPiVBOX:~# journalctl -u dietpi-vpn --no-pager
Mar 15 12:33:23 DietPiVBOX systemd[1]: Starting VPN Client (DietPi)...
Mar 15 12:33:23 DietPiVBOX openvpn[1799]: OpenVPN 2.4.7 x86_64-pc-linux-gnu [SSL (OpenSSL)] [LZO] [LZ4] [EPOLL] [PKCS11] [MH/PKTINFO] [AEAD] built on Feb 20 2019
Mar 15 12:33:23 DietPiVBOX openvpn[1799]: library versions: OpenSSL 1.1.1d  10 Sep 2019, LZO 2.10
Mar 15 12:33:23 DietPiVBOX systemd[1]: Started VPN Client (DietPi).
Mar 15 12:33:23 DietPiVBOX openvpn[1799]: NOTE: the current --script-security setting may allow this configuration to call user-defined scripts
Mar 15 12:33:23 DietPiVBOX openvpn[1799]: Outgoing Control Channel Authentication: Using 512 bit message hash 'SHA512' for HMAC authentication
Mar 15 12:33:23 DietPiVBOX openvpn[1799]: Incoming Control Channel Authentication: Using 512 bit message hash 'SHA512' for HMAC authentication
Mar 15 12:33:23 DietPiVBOX openvpn[1799]: TCP/UDP: Preserving recently used remote address: [AF_INET]89.39.107.190:80
Mar 15 12:33:23 DietPiVBOX openvpn[1799]: Socket Buffers: R=[212992->212992] S=[212992->212992]
Mar 15 12:33:23 DietPiVBOX openvpn[1799]: UDP link local: (not bound)
Mar 15 12:33:23 DietPiVBOX openvpn[1799]: UDP link remote: [AF_INET]89.39.107.190:80
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: TLS: Initial packet from [AF_INET]89.39.107.190:80, sid=65720a3b 06535dec
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: WARNING: this configuration may cache passwords in memory -- use the auth-nocache option to prevent this
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: VERIFY OK: depth=2, C=CH, O=ProtonVPN AG, CN=ProtonVPN Root CA
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: VERIFY OK: depth=1, C=CH, O=ProtonVPN AG, CN=ProtonVPN Intermediate CA 1
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: VERIFY KU OK
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: Validating certificate extended key usage
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: ++ Certificate has EKU (str) TLS Web Client Authentication, expects TLS Web Server Authentication
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: ++ Certificate has EKU (oid) 1.3.6.1.5.5.7.3.2, expects TLS Web Server Authentication
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: ++ Certificate has EKU (str) TLS Web Server Authentication, expects TLS Web Server Authentication
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: VERIFY EKU OK
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: VERIFY OK: depth=0, CN=nl-117.protonvpn.com
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: Control Channel: TLSv1.3, cipher TLSv1.3 TLS_AES_256_GCM_SHA384, 4096 bit RSA
Mar 15 12:33:24 DietPiVBOX openvpn[1799]: [nl-117.protonvpn.com] Peer Connection Initiated with [AF_INET]89.39.107.190:80
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: SENT CONTROL [nl-117.protonvpn.com]: 'PUSH_REQUEST' (status=1)
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: PUSH: Received control message: 'PUSH_REPLY,dhcp-option DNS 10.19.0.1,redirect-gateway def1,sndbuf 524288,rcvbuf 524288,explicit-exit-notify,comp-lzo no,route-gateway 10.19.0.1,topology subnet,ping 10,ping-restart 60,ifconfig 10.19.0.12 255.255.0.0,peer-id 196626,cipher AES-256-GCM'
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: timers and/or timeouts modified
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: explicit notify parm(s) modified
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: compression parms modified
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: --sndbuf/--rcvbuf options modified
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: Socket Buffers: R=[212992->425984] S=[212992->425984]
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: --ifconfig/up options modified
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: route options modified
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: route-related options modified
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: --ip-win32 and/or --dhcp-option options modified
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: peer-id set
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: adjusting link_mtu to 1657
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: OPTIONS IMPORT: data channel crypto options modified
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: Data Channel: using negotiated cipher 'AES-256-GCM'
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: Outgoing Data Channel: Cipher 'AES-256-GCM' initialized with 256 bit key
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: Incoming Data Channel: Cipher 'AES-256-GCM' initialized with 256 bit key
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: ROUTE_GATEWAY 192.168.1.1/255.255.255.0 IFACE=eth0 HWADDR=08:00:27:fe:c9:14
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: TUN/TAP device tun0 opened
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: TUN/TAP TX queue length set to 100
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: /sbin/ip link set dev tun0 up mtu 1500
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: /sbin/ip addr add dev tun0 10.19.0.12/16 broadcast 10.19.255.255
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: /sbin/ip route add 89.39.107.190/32 via 192.168.1.1
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: /sbin/ip route add 0.0.0.0/1 via 10.19.0.1
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: /sbin/ip route add 128.0.0.0/1 via 10.19.0.1
Mar 15 12:33:25 DietPiVBOX openvpn[1799]: Initialization Sequence Completed
```

**Comment by MichaIng** [comment]
```
0.0.0.0/1 via 10.19.0.1 dev tun0 
128.0.0.0/1 via 10.19.0.1 dev tun0 
```
Looks good. `tcpdump` could be used to check where packages are in fact routed.

**Comment by nonnorm** [comment]
Okay, so just running `curl https://ifconfig.co` with no parameters returns my IP(v6) address.  Running `curl -6` returns an error.  Running `curl -4` returns the VPN server's address.  One more interesting thing is that when running a `tcpdump` on one of the ports used by the server, it actually shows that it is pinging the VPN server every few seconds depending on the server:
```text
root@DietPiVBOX:~# tcpdump port 4569
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), capture size 262144 bytes
15:02:40.169953 IP 217.138.206.83.4569 > DietPiVBOX.lan.49598: UDP, length 41
15:02:40.170054 IP DietPiVBOX.lan.49598 > 217.138.206.83.4569: UDP, length 41
15:02:50.189143 IP DietPiVBOX.lan.49598 > 217.138.206.83.4569: UDP, length 41
- 15:02:50.268341 IP 217.138.206.83.4569 > DietPiVBOX.lan.49598: UDP, length 41
```
Anyway, it seems to probably be something with my network setup and/or this VM being linked with my computer, and not a problem with the implementation.  Thus it's ready on my end.

**Comment by MichaIng** [comment]
Ah, so it's IPv4 vs IPv6. Makes sense, since IPv6 requires own routes. What are they?
```
ip -6 r
```

There is a `redirect-gateway` which defines which routes are created by default. Actually quite important to rule out that IPv6 requests bypass the VPN, and it IPv6 is not supported server-side, is needs to be disabled on the client. We can implement a test and theoretically even add up/down rules to disable/re-enable IPv6 on connection start/stop, or permanently.

**Comment by nonnorm** [comment]
```
root@DietPiVBOX:~# ip -6 r
::1 dev lo proto kernel metric 256 pref medium
2600:6c56:4100:1317::/64 dev eth0 proto kernel metric 256 expires 259167sec pref medium
fe80::/64 dev eth0 proto kernel metric 256 pref medium
fe80::/64 dev tun0 proto kernel metric 256 pref medium
default via fe80::366b:46ff:fe30:5c5a dev eth0 proto ra metric 1024 expires 8947sec pref medium
```
Looks like the `eth0` rules take precedence over the `tun0` rules?  Regardless, you're right, the servers (that I tested, at least) don't support IPv6, as verified on my main computer using the native client.
```
~ » curl -6 https://ifconfig.co
108.59.0.40
```

**Comment by MichaIng** [comment]
There is no IPv6 VPN route, only one for link-local (LAN) addresses but none that would cover the whole IP range. Does `redirect-gateway ipv6` in the client config force those routes to be created?

**Comment by nonnorm** [comment]
Whoops, thought the `fe80::/64` rules said `tun0`.  No, adding that to the config file changes nothing.

**Comment by MichaIng** [comment]
`fe80::/64` is the local-link block, so you can compare this roughly with `10.19.0.0/16` on IPv4. Not sure why it's no more specific in case of IPv6, since now there are two identical routes for IPv6 LAN requests 🤔.

Okay, however, looks like at least with ProtonVPN we need to disable IPv6 either permanently or temporarily. I'll check the behaviour with NordVPN.

**Comment by nonnorm** [comment]
I just noticed, the ProtonVPN CLI also disables IPv6 when connecting.  https://github.com/ProtonVPN/linux-cli/blob/089af8d309cf774adc31c158a62a27733ae39ee0/protonvpn_cli/connection.py#L501.  Also, should we use up/down scripts for this, to keep it working even on autostart?

**Comment by Joulinar** [comment]
Should we inform users before going to disable IPv6? Are there users on an only IPv6 environment?

**Comment by MichaIng** [comment]
AN only IPv6 environment would not work anyway, but of theoretically the LAN network could be IPv6 indeed. We can add a prompt when ProtonVPN is selected + IPv6 currently enabled, that it will be disabled during VPN connection.

Yes up/down options can be used. For the options we use route-up/route-pre-down, so both when the VPN connection is up, instead of before its up and after its down.
The following commands will do:
```
/boot/dietpi/func/dietpi-set_software enableipv6 1
/boot/dietpi/func/dietpi-set_software enableipv6 0
```

**Comment by nonnorm** [comment]
Wouldn't it actually be better to use `ip6tables`,  as if people turned on IPv6 later it would still work?

**Comment by nonnorm** [comment]
Also, thoughts on removing the `misc` folder, moving `dietpi-vpn` and `dietpi-justboom` to the main folder, and all the rest to the `func` folder?

**Comment by nonnorm** [comment]
Okay, it's ready on my end.

File: dietpi/misc/dietpi-vpn [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+	VPN_USERNAME=
+	VPN_PASSWORD=
+	VPN_SERVER=
+	PROTOCOL='udp'
+	VPN_CONNECTED=0
+	WAN_IP=
+	RX='N/A'
+	TX='N/A'
+
+	Update_Wan_Ip(){ WAN_IP=$(curl -4sSfLm 2 https://ifconfig.co/json 2>&1 | mawk -F\" '/"ip"/{o=$4}/"country"/{c=$4}/"region_name"/{c=$4"-"c}END{if(c){o=o" "c}if(o){print o;exit}{print}}'); }
```
Comment (MichaIng) [inline dietpi/misc/dietpi-vpn]: Is it a specific reason why IPv4 is forced for the IP? I think it's fine to show the IPv6 address when this is what is used by default.

**Comment by MichaIng** [comment]
Awesome that you added a killswitch! 😍

Further suggestions:
- Let's not use $FP_SETTINGS_UP/$FP_SETTINGS_DOWN for disabling IPv6 and the killswitch. Since they are implemented via `route-up` and `route-pre-down`, they are basically called them the VPN is still up, and also it messes with custom user scripts too easily. I think `up`/`down` can be used multiple times, so we could even implement two additional script, one for IPv6 (statically added via .ovpn template) and one dynamically added for the killswitch.
- I'd like to have a provider selection menu. So one can change it easily. But of course at least the server menu would need to be hidden until a provider has been selected. That menu needs to be hidden anyway, when a custom .ovpn file is used.
- ... I'll run further tests with NordVPN this evening.

**Comment by nonnorm** [comment]
It should ask for a provider on first run if nothing is installed already (though my implementation is kind of hacky, there's probably a better way to do it).

**Comment by MichaIng** [comment]
I agree, the provider menu should be shown automatically when none was selected/installed already, e.g. if neither sub directory exist, nor a custom client.ovpn.

**Comment by nonnorm** [comment]
That's actually what it does (double meaning of 'should').

https://github.com/MichaIng/DietPi/blob/c65780a7212720ed867b2ab83d7bcf2cf35282ad/dietpi/misc/dietpi-vpn#L465

**Comment by MichaIng** [comment]
I see the update and install parts being in one function. So an individual provider selection even when one has been selected before, requires those to be separated. Not too important for now, I'll add it later when finding time.

**Comment by MichaIng** [comment]
Added the provider menu and separated the killswitch scripts from the custom up/down scripts. However, the killswitch does not yet work as the syntax of the rules file seems to be wrong:
```
*filter
:INPUT DROP [0:0]
:FORWARD DROP [0:0]
:OUTPUT DROP [0:0]
iptables -P INPUT DROP
iptables -P FORWARD DROP
iptables -P OUTPUT DROP
iptables -A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
iptables -A OUTPUT -o lo -j ACCEPT
iptables -A OUTPUT -o tun0 -p icmp -j ACCEPT
iptables -A OUTPUT -d 192.168.0.0/16 -j ACCEPT
iptables -A OUTPUT -d 172.16.0.0/12 -j ACCEPT
iptables -A OUTPUT -d 10.0.0.0/8 -j ACCEPT
iptables -A OUTPUT -p tcp -m multiport --dports 80,443,1194,5060,4569 -d $VPN_SERVER -j ACCEPT
iptables -A OUTPUT -p udp -o -m multiport --dports 80,443,1194,5060,4569 -d $VPN_SERVER -j ACCEPT
iptables -A OUTPUT -o tun0 -j ACCEPT
COMMIT
```
I guess calling `iptables` within the rules file is wrong here?
...
Jep, I just applied the commands manually, fixed the "-o -m" which should be "-m" only, and iptables-save shows what we want:
```
*filter
:INPUT DROP [0:0]
:FORWARD DROP [0:0]
:OUTPUT DROP [0:0]
-A INPUT -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A OUTPUT -o lo -j ACCEPT
-A OUTPUT -o tun0 -p 1 -j ACCEPT
-A OUTPUT -d 192.168.0.0/16 -j ACCEPT
-A OUTPUT -d 172.16.0.0/12 -j ACCEPT
-A OUTPUT -d 10.0.0.0/8 -j ACCEPT
-A OUTPUT -d $VPN_SERVER -p 6 -m multiport --dports 80,443,1194,5060,4569 -j ACCEPT
-A OUTPUT -d $VPN_SERVER -p 17 -m multiport --dports 80,443,1194,5060,4569 -j ACCEPT
-A OUTPUT -o tun0 -j ACCEPT
COMMIT
```

**Comment by MichaIng** [comment]
`-A OUTPUT -o tun0 -p icmp -j ACCEPT` is obsolete when `-A OUTPUT -o tun0 -j ACCEPT` is given, isn't it?

**Comment by nonnorm** [comment]
Yes, it is.  Sorry, I compiled that list from a few sources.  I guess I missed that duplicate entry.

**Comment by MichaIng** [comment]
NordVPN btw also does not support IPv6. So I'd say we should always disable it when connecting to the VPN. I suggest we use static up/down scripts, where IPv6 is disabled on system level (via sysctl) when connection is established and depending on dietpi.txt re-enabled on disconnect. The killswitch then needs to be appended/removed from those scripts, as it's not possible to use two up or down keys in the config file: The second setting overrides the first, i.e. only one of them is executed.

**EDIT: Easier is that the scripts automatically enable/disable the killswitch if the files exist. So we only need to create/remove the files and do not need to alter any files.**

**Comment by nonnorm** [comment]
Well, now it doesn't even look like the script is being run.  It doesn't turn off IPv6, and it can't create a file in my home directory.

**Comment by MichaIng** [comment]
> Well, now it doesn't even look like the script is being run. It doesn't turn off IPv6, and it can't create a file in my home directory.

- There was a syntax when applying the killswitch rules. But it should not affect disabling IPv6, let me see. Probably it's easier to call the scripts via ExecStartPre and ExecStopPost commands from the systemd service. It's not strictly tied to the actual OpenVPN connection itself (which can be started manually) but to the service only, but further simplifies the client configurations and allows even custom up/down commands (instead of only route-up and route-pre-down) 🤔.
- It's expected that the service cannot write to any home directory due to `ProtectHome=true` in the systemd unit. It's a system service, not a user service, so must not be setup to read/write anything to/from a login users home directory, at least that's the idea behind those (default) service hardenings.
____________
Disabling and re-enabling IPv6 works here as expected:
```
Mar 28 16:51:41 VM-Buster openvpn[2369]: /var/lib/dietpi/dietpi-vpn/static_up.sh tun0 1500 1585 10.8.2.3 255.255.255.0 init
Mar 28 16:51:41 VM-Buster openvpn[2369]: net.ipv6.conf.all.disable_ipv6 = 1
Mar 28 16:51:41 VM-Buster openvpn[2369]: net.ipv6.conf.default.disable_ipv6 = 1
Mar 28 16:51:42 VM-Buster openvpn[2369]: /sbin/ip route add 185.220.70.235/32 via 192.168.1.1
Mar 28 16:51:42 VM-Buster openvpn[2369]: /sbin/ip route add 0.0.0.0/1 via 10.8.2.1
Mar 28 16:51:42 VM-Buster openvpn[2369]: /sbin/ip route add 128.0.0.0/1 via 10.8.2.1
Mar 28 16:51:42 VM-Buster openvpn[2369]: Initialization Sequence Completed
Mar 28 16:52:27 VM-Buster openvpn[2369]: event_wait : Interrupted system call (code=4)
Mar 28 16:52:27 VM-Buster openvpn[2369]: SIGTERM received, sending exit notification to peer
Mar 28 16:52:27 VM-Buster systemd[1]: Stopping VPN Client (DietPi)...
Mar 28 16:52:28 VM-Buster openvpn[2369]: /sbin/ip route del 185.220.70.235/32
Mar 28 16:52:28 VM-Buster openvpn[2369]: /sbin/ip route del 0.0.0.0/1
Mar 28 16:52:28 VM-Buster openvpn[2369]: /sbin/ip route del 128.0.0.0/1
Mar 28 16:52:28 VM-Buster openvpn[2369]: Closing TUN/TAP interface
Mar 28 16:52:28 VM-Buster openvpn[2369]: /sbin/ip addr del dev tun0 10.8.2.3/24
Mar 28 16:52:28 VM-Buster openvpn[2369]: /var/lib/dietpi/dietpi-vpn/static_down.sh tun0 1500 1585 10.8.2.3 255.255.255.0 init
Mar 28 16:52:28 VM-Buster openvpn[2369]: net.ipv6.conf.all.disable_ipv6 = 0
Mar 28 16:52:28 VM-Buster openvpn[2369]: net.ipv6.conf.default.disable_ipv6 = 0
Mar 28 16:52:28 VM-Buster openvpn[2369]: SIGTERM[soft,exit-with-notification] received, process exiting
Mar 28 16:52:28 VM-Buster systemd[1]: dietpi-vpn.service: Succeeded.
Mar 28 16:52:28 VM-Buster systemd[1]: Stopped VPN Client (DietPi).
```

**Review by MichaIng — DISMISSED** [review]

**Review by MichaIng — DISMISSED** [review]

**Review by MichaIng — APPROVED** [review]

**Comment by MichaIng** [comment]
@ravenclaw900 
Does it work for you now with switching IPv6?

**Comment by nonnorm** [comment]
Screen Shot 2021-03-28 at 2 16 48 PM
Looks like it.

**Comment by MichaIng** [comment]
Perfect, let's merge.

**Comment by zizou1912** [comment]
> **Status**: Ready
> 
> * [x]  Add way to connect to ProtonVPN
> * [x]  Add way to just use generic OpenVPN file
> * [x]  Test with ProtonVPN
> * [x]  Test with NordVPN
> * [x]  Add killswitch
> 
> **Reference**: #3948
> 
> **Commit list/description**:
> 
> * DietPi-VPN | Start transition from dietpi-nordvpn to dietpi-vpn
> * DietPi-Software | Add way to download and parse ProtonVPN server list
> * DietPi-VPN | Add way to select ProtonVPN Servers
> * DietPi-VPN | Add way to generate .ovpn file
> * DietPi-Software | Switch from NordVPN to VPN
> * DietPi.bash | Add missing DietPi-VPN alias
> * Patches | Remove DietPi-NordVPN

Could you please add support for "Private Internet Access" Wireguard VPN? They have an open-source script for it. Here https://www.privateinternetaccess.com/helpdesk/kb/articles/manual-connection-and-port-forwarding-scripts

**Comment by nonnorm** [comment]
Hello,
Note that you can use any `.ovpn` file with DietPi-VPN, and installing WireGuard from DietPi-Software will let you use it as a client.  DietPi-VPN doesn't support custom VPN starting scripts.

**Comment by zizou1912** [comment]
Ture, but wireguard setup is complicated using their script. Idont know
(they don't provide wireguard config, but rather a script that generate
them every day).

I thought it would be nice to integrate their script if u already have
integrated custom Nordvpn Wireguard and others.🙂 Just an idea.

Openvpn utilizing single core is so much slower. 260Mbps vs 780Mbps.

**Comment by nonnorm** [comment]
As of yet, DietPi-VPN doesn't actually support WireGuard, just OpenVPN.

**Comment by MichaIng** [comment]
Indeed, WireGuard for public VPN providers is difficult, as WireGuard itself requires static IPs for each client, stored in the server config, which is a privacy issue. For this reason, public providers do not use WireGuard natively, but wrap it into own custom scripts or tools, often but not always open-source, to have a wrapping NAT or regular IP rotation.

Implementing/copying and maintaining/updating the needs for those WireGuard wrappers for each VPN provider into DietPi-VPN is IMO not reasonable. If someone finds time to figure it out for a particular provider, and it does not explode the size of the script, I'm fine with it, but from myself it would draw too much time away from more important things.

Light at the horizon is that the WireGuard developers are working on an internal dynamic IP implementation already. Not sure what state it is, but with that it should be possible for providers to ship a native WireGuard client config file, which can then be reasonable natively integrated into DietPi-VPN 😃.
- Official idea: https://github.com/WireGuard/wg-dynamic/blob/master/docs/idea.md
- Last commit, 8 months ago 🤔: https://github.com/WireGuard/wg-dynamic/branches/all

**Comment by zizou1912** [comment]
Yeah that's totally understandable. I thought at the beginning you guys did
some sort of such implementation with Nordvpn.

And thanks for the great news. I didn't know they're working on dynamic
IPs. Jeez i don't know how they slept through it from the start😂. Probably
for the promo (oNlY 4000 LiNeS oF cOdE).

Edit: Oh, Yiiikes, seems pretty dead, wonder why they're so uninterested in it.

**Comment by Joulinar** [comment]
@ravenclaw900 
@MichaIng 
can someone of you have a look to the forum https://dietpi.com/phpbb/viewtopic.php?t=8997

There is a user having issues using IPv6 while NordVPN is enabled. If I understood correctly, it has been disabled by design https://github.com/MichaIng/DietPi/pull/4180#issuecomment-808753586

**Comment by MichaIng** [comment]
> If I understood correctly, it has been disabled by design

Exactly. Otherwise IPv6 requests would bypass the VPN. NordVPN and ProtonVPN both do not support IPv6.

**Pull request merged**